Added MineClone support

ApolloX939 opened the pull request:
This PR contains a little bit of work:
- Moved recipes/crafting to support Minetest (default) and MineClone (mcl_core).
- Made copies of Minetest Game's chest's textures (this allows us to use textures without loosing them based on Minetest or MineClone)
- Modified Shared Chest to support MineClone's groups for nodes (we mine the shared chest in the same speed a MineClone's chest)
- Changed the HUD so it will show green for access in a claim, it will be red if you don't have access but that area has been claimed, and white in areas that are not claimed (though if there is no claim you'll hide it anyway so it really will only show green or red).

All except the HUD color change has been tested on local copies of Minetest Game and MineClone2 (MineClone5 also needs to be tested, since it incorperates mcl_copper we'd actually use copper instead of falling back to gold).

Commit:
Added MineClone support, also changed HUD

  I've changed HUD so it will be green if you have permission/access,
red if you don't and white/grey if it's unclaimed. (this needs areas to
be present to show the hud)

Commit:
Updated MineClone Shared chest

  Yay, you can now successfully use the Shared Chest with MineClone!
(And pick it back up!)

SmallJoker commented on chest.lua:
```diff
@@ -11,35 +11,69 @@ local function get_item_count(pos, player, count)
+-- Just in case we are under MineClone
+local default = rawget(_G, "default") or nil
+if default == nil then
+	default = rawget(_G, "mcl_sounds")
```
Please add local variables to specify `default.node_sound_wood_defaults` game-dependent configuration. This line relies on 100% API of `default` and `mcl_sounds`, which is by far not true.

Suggestion: use something similar to pipeworks: https://github.com/mt-mods/pipeworks/blob/master/crafts.lua#L1

ApolloX939 replied:
I don't quite understand what your pointing out with this.

The code makes it so if default doesn't exist our usage of `default.node_sound_wood_defaults()` is the exact same, with 0 difference code wise. (Unless we use other things in default this works just fine, but I guess I could take the output from it as a variable and change based on availability)

ApolloX939 replied:
Like [https://github.com/Beanzilla/item_replicator/blob/master/register.lua#L314-L324](https://github.com/Beanzilla/item_replicator/blob/master/register.lua#L314-L324)?

SmallJoker replied:
Yes, the code stays the same, but `default` it not guaranteed to support all MTG API that's expected from it (if using MineClone). Even though it does not result in errors now, it's a hack that is better implemented in the manner of the mod you linked above.

ApolloX939 replied:
Yeah I've reworked this to like my mod, we use the one feature we need from it and not try to load it in it's entirety.

SmallJoker commented on chest.lua:
```diff
@@ -11,35 +11,69 @@ local function get_item_count(pos, player, count)
+		"simple_protection_chest_side.png" .. tex_mod,
+		"simple_protection_chest_side.png" .. tex_mod,
+		"simple_protection_chest_side.png" .. tex_mod,
+		"simple_protection_chest_lock.png" .. tex_mod
```
You could use the mcl chest texture for this, if it's compatible.

ApolloX939 replied:
Mcl chest is a entity, so not sure if we can. (That's why I just simply ported over internal copies)

SmallJoker replied:
I see. This approach makes sense in that case.

ApolloX939 replied:
Anywho, changed this so it is using the chest textures. (Our chest may look a little different but for the most part it should look about the same, slightly bigger as we are a node not an entity but so long as we don't try depicting someone opening the chest thus having an opened state we should be fine)

SmallJoker commented on chest.lua:
```diff
@@ -11,35 +11,69 @@ local function get_item_count(pos, player, count)
+		-- Minetest Game
+		choppy = 2, oddly_breakable_by_hand = 2,
+		-- MineClone
+		handy=1,axey=1, deco_block=1,
```
Should be moved to a variable which changes depending on the game.

ApolloX939 replied:
Technically speaking it can be left here, since Minetest Game uses different variables/groups than MineClone so there is no collision, but if you insist on it being a var okay. (I don't care so long as it works)

SmallJoker replied:
Right, I guess it does not matter too much for node groups. In the worst case, a group could be misinterpreted as another functionality/property that is not intended by one of the games.

ApolloX939 replied:
Anyway I've move this also to a variable changing it depending on game.

SmallJoker commented on chest.lua:
```diff
@@ -11,35 +11,69 @@ local function get_item_count(pos, player, count)
+			)
+			local inv = meta:get_inventory()
+			inv:set_size("main", 8*4)
+		elseif minetest.registered_items["mcl_core:dirt"] then
```
This should definitely not depend on random items. Moving this function to a local variable (same as suggested before) on a per-game basis would be cleaner.

SmallJoker commented on chest.lua:
```diff
@@ -11,35 +11,69 @@ local function get_item_count(pos, player, count)
+			local mcl_formspec = rawget(_G, "mcl_formspec") or nil
+			meta:set_string("formspec",
+				"size[9,8.75]"..
+				"label[0,0;"..minetest.formspec_escape(minetest.colorize("#313131", "Shared Chest")).."]"..
```
`S("Shared Chest")`

ApolloX939 replied:
Oops, missed the translator.

SmallJoker commented on chest.lua:
```diff
@@ -11,35 +11,69 @@ local function get_item_count(pos, player, count)
+		elseif minetest.registered_items["mcl_core:dirt"] then
+			-- In MineClone they have a separate overlay of images for the inventory,
+			-- This branch should setup the proper dimentions (9x3) instead of Minetest's fairly nice dimentions (8x4)
+			local mcl_formspec = rawget(_G, "mcl_formspec") or nil
```
Access `mcl_formspec` directly and whitelist the variable in `.luacheckrc`. Things would error anyway if this were missing in MineClone 2.

SmallJoker commented on chest.lua:
```diff
@@ -68,14 +102,29 @@ minetest.register_node("simple_protection:chest", {
-	recipe = { "simple_protection:claim", "default:chest_locked" }
-})
+-- If neither of these occur then the shared chest is just uncraftable
+if minetest.registered_items["default:dirt"] then
```
Use a variable instead. Check against the mod presence, and not random items.

SmallJoker commented on hud.lua:
```diff
@@ -25,11 +25,13 @@ if minetest.get_modpath("areas") then
-	local color = 0xFFFFFF
-	if has_access then
+local function generate_hud(player, current_owner, has_access, claimed)
+	-- green if access, red if no access but claimed, grey for unclaimed
```
Unrelated change. Please move this to a separate PR.

( *The text for unclaimed areas (`"Area owner: `) could be better, such as `Area is not claimed`.* )

ApolloX939 replied:
Yeah I need to pull those changes out so I can make a separate PR. (But first let's get the MineClone support stuff going)

SmallJoker commented on hud.lua:
```diff
@@ -60,8 +63,10 @@ minetest.register_globalstep(function(dtime)
 
 		local current_owner = ""
 		local data = sp.get_claim(player:get_pos())
+		local claimed = false
```
Variable is not needed. You can re-use the `current_owner`/`owner` field.

ApolloX939 replied:
That I didn't know... Today I Learned (TIL)

SmallJoker commented on protection.lua:
```diff
@@ -90,14 +90,39 @@ minetest.register_craftitem("simple_protection:claim", {
-	}
-})
+-- If neither one of these occurs then the claim stick isn't craftable
+if minetest.registered_items["default:dirt"] then
```
Use mod names to distinguish games. Perhaps move such variable into `init.lua` for mod-wide access?

ApolloX939 replied:
Yup, where it should be.

SmallJoker commented on protection.lua:
```diff
@@ -90,14 +90,39 @@ minetest.register_craftitem("simple_protection:claim", {
+		}
+	})
+elseif minetest.registered_items["mcl_core:dirt"] then
+	if minetest.registered_items["mcl_copper:stone_with_copper"] then
```
This check is fine (the material is actually related), but consider using variables like demonstrated in pipeworks.

ApolloX939 replied:
Yeah didn't know how high tech you wanted me to make this mod. (I'll redo it with as best I know)

SmallJoker replied:
I see that it takes some effort to get it there, but afterwards it will be easy to extend to support more games.

ApolloX939 replied:
Yup, just need to work on getting it to that spot in the most reliable manner.

SmallJoker commented:
Thank you for this PR. I actually had the same idea yesterday, but postponed it.

Commit:
Reversed HUD change, focusing on MineClone

  This should reverse all the changes done to the HUD.

  Now to work on refactoring MineClone support

Commit:
Unified game mode detection

  This would be the first spot to change for adding another game support

Commit:
Unified itemstrings in crafting

  This would be your 2nd spot to change to add support for another game.

Commit:
Unified Chest and Claim usage

  This consolodated some of the codebase for multi-game support (so
you'd just update there)

Commit:
Added MCL textures for chest in MineClone

  We now don't need to keep our own textures around.

ApolloX939 commented:
Ok, so looks like my rewrite/refactor is almost done... if not done. :smiley:

ApolloX939 commented:
Yeah I add a few quirky new optional_depends (mcl_core, mcl_formspec, mcl_chests (just for the textures), etc).

But this should allow us to get multi-game support... and it's to some degree easy to change/update for more games. (I hope I made it simpler)

SmallJoker commented:
Squashed & merged as f38eba5e6a

Thank you for your contribution and the changes.
